### DISCUSSION TITLE
Inner block supports and future plans

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#35473] [Developer Experience]
Maybe I missed this and it's already there but, are there plans to let blocks decide where they want to apply the CSS of some of the blocks supports?

For example, what if I want my block to support a background color but it should not be applied to the root HTML node, but instead to an inner node?

<img src="https://user-images.githubusercontent.com/3305402/136599073-d570d38d-1460-433b-b480-4b3477e2d60c.png" width="300" />

Also:
- Are there any plans to use Block Supports for things that require, not only CSS but HTML modifications? _Again, maybe this already exists and I missed it._
- Are there any plans to use Block Supports for things that would require JS hydration in the client?

**Comment by mtias** [comment]
> For example, what if I want my block to support a background color but it should not be applied to the root HTML node, but instead to an inner node?

Yes, it was scoped mostly in this issue: https://github.com/WordPress/gutenberg/issues/28913.

**Reply by luisherranz** [reply]
Thanks a lot for the response Matías, and sorry for my late answer.

I've been taking a look at #28913. It's very interesting, thanks. It seems like it can provide more flexibility in that aspect. I still feel that I need to explore block supports more to understand the full scope of their usage and current design.

I've also stumbled upon this other @oandregal issue (#22887) that highlights the need to repeat block supports and contexts as block attributes to add typings and defaults, which is also interesting.

I wonder if we could find a way in the future to declare the attributes/supports/contexts directly in the DOM/template to solve this problem, but I don't have any specific idea right now.

By the way, regarding my questions about supports that require JS or HTML changes: Is this something that has been explored before, or is it open for exploration? 🙂

**Reply by mtias** [reply]
The thing with declaring them in the DOM/template is that it blurs the line between the block instance and the block type if it initializes with default values or serializes them.

One of the motivations for supports is to decouple them from markup as much as possible, so that they can be assembled as needed at any point. Requiring explicit markup changes to accommodate side effects would be tricky in that regard.

**Reply by luisherranz** [reply]
As I said, I don't have any specific idea about this yet, so I'm just exploring ideas... but I feel that an in-place markup declaration in some sort of template system could have some potential 🙂

For example, a button block could become something like this (block supports are implicitly declared and don't need to be repeated elsewhere):

```html
<div @supports:color.background={$backgroundColor} @supports:fontSize>
  <a @supports:color.text @supports:border.radius="25px" @supports:spacing>
    {$attributes["content"]}
  <a/>
</div>
```

_There are many many ways to design this, this is just one._

This would later be compiled to whatever code is necessary for WordPress to know what block supports are declared for this block and what HTML node they should be applied to.

> The thing with declaring them in the DOM/template is that it blurs the line between the block instance and the block type if it initializes with default values or serializes them

What problem would you see regarding initialization in that example?

> decouple them from markup as much as possible, so that they can be assembled as needed at any point. 
> Requiring explicit markup changes to accommodate side effects would be tricky in that regard

Could you share an example of that so I can understand the limitation?

**Reply by luisherranz** [reply]
> Are there any plans to use Block Supports for things that require, not only CSS but HTML modifications?

Answering myself here, but I've been trying to think about this question and I cannot find a specific example of a styling capability that requires additional markup (or a modification of it). Maybe something that involves `flex` or `grid`...?

**Comment by luisherranz** [comment]
By the way, I think this @c4rl0sbr4v0's discussion is related to our conversation here: https://github.com/WordPress/gutenberg/discussions/36079

**Reply by gziolo** [reply]
#36079 might be related when you want to apply class names or styles to the inner blocks. However, in the majority of cases, the issue is that block supports generate attributes only for the block wrapper when in some cases you might need to use some of them with nested HTML tags of the block. @c4rl0sbr4v0 used `useBlockProps` and it's matching util in PHP with the child of the block's root HTML tag and while it applies all styles correctly, it makes the block selection behave differently. Ideally, we make the API flexible enough so you could distribute block supports features more freely between HTML markup of the block.

**Reply by luisherranz** [reply]
> Ideally, we make the API flexible enough so you could distribute block supports features more freely between HTML markup of the block

Yes! I opened this discussion to learn more about this. Thanks Greg 🙂

I had a bit of a hard time understanding the final API proposed in #28913 to solve this problem because it's scattered in many different issues/PRs.

Does it solve the #36079 issue? Because I have the impression that it doesn't.

**Reply by gziolo** [reply]
> Does it solve the #36079 issue? Because I have the impression that it doesn't.

@c4rl0sbr4v0 might know better the answer. As far as I can tell we would have to try using the experimental flag `__experimentalSkipSerialization` like in the Button block:

https://github.com/WordPress/gutenberg/blob/b55cede14c3a6f0ed797d73e676c9fc504fd58d4/packages/block-library/src/button/block.json#L60-L61

https://github.com/WordPress/gutenberg/blob/b55cede14c3a6f0ed797d73e676c9fc504fd58d4/packages/block-library/src/button/block.json#L69-L70

https://github.com/WordPress/gutenberg/blob/b55cede14c3a6f0ed797d73e676c9fc504fd58d4/packages/block-library/src/button/block.json#L76-L78

There are corresponding hooks that let you read those skipped features:

https://github.com/WordPress/gutenberg/blob/b55cede14c3a6f0ed797d73e676c9fc504fd58d4/packages/block-library/src/button/edit.js#L123-L125

They need to be mapped to class names and styles:

https://github.com/WordPress/gutenberg/blob/b55cede14c3a6f0ed797d73e676c9fc504fd58d4/packages/block-library/src/button/edit.js#L170-L185

I'm not sure how it could be translated to code in `render_callback` though.

That's way more complex than what you need to do when all features get converted to styles and classes for the block wrapper.

**Reply by luisherranz** [reply]
That's much clearer now. Thanks Greg.

It'd be awesome if you could add this API explanation to the #28913 opening post for the newcomers. To be honest, it is almost impossible to figure it out by jumping around and reading code from all the issues and PRs involved in #28913 🙂🙏

> I'm not sure how it could be translated to code in render_callback though.

@c4rl0sbr4v0 if you need to spread the supports attributes in the `render_callback` function as well, could you please report how you did it here as well?

**Reply by cbravobernal** [reply]
Right now if you use the default function `get_block_wrapper_attributes()` returns a string with all set attributes like this one:
`style="border-radius: 100px" class="has-background-color olive-colour`

Taking a look at that function, you can check that at the end, implodes an array of those attributes, so the previous function call:
`WP_Block_Supports::get_instance()->apply_block_supports();`, returns an array with separated styles and classes.

But still, there is not an easy way to get attributes spread in case you need to separate styles or separate classes. Maybe we could investigate if we can return all styles in an array instead of a string, that way we could apply them wherever we need.

Could be a cool issue to investigate-develop in the short future, maybe the comment loop template or even the image block could be a good option to start with (so we can achieve the background style proposed in #36079)

**Reply by luisherranz** [reply]
> WP_Block_Supports::get_instance()->apply_block_supports();, returns an array with separated styles and classes.

Can you also access them separated by block support (like border, spacing, color...) so you can put each one in the correct HTML node?

**Reply by gziolo** [reply]
> Can you also access them separated by block support (like border, spacing, color...) so you can put each one in the correct HTML node?

I don't think it's possible to access them by type. It's worth noting, that colors can be multiple settings like text, link, background or border (color). One idea I shared in a completely different discussion that I can't find now would be to provide scoping for block supports. By default, it would be the block wrapper, but you could define some custom scopes so in PHP you would have:

```php
<?PHP

$wrapper_attributes = get_block_wrapper_attributes();
$scoped_attributes  = get_block_scope_attributes( 'my-scope' );

return sprintf(
    '<div %1$s><p %2$s>Some content</p></div>',
    $wrapper_attributes,
    $scoped_attributes
);
```

A related code in `block.json` could look as follows:

```json
"supports": {
    "color": {
        "scope": "my-scope"
    }
}
```

**Reply by gziolo** [reply]
I left a comment with another proposal in https://github.com/WordPress/gutenberg/discussions/36079#discussioncomment-2033349. I'm wondering now whether I should move it here and we should consolidate the discussion in this place.

**Reply by luisherranz** [reply]
Why not in https://github.com/WordPress/gutenberg/discussions/37495?

**Reply by gziolo** [reply]
That’s another option. I’m just trying to to figure out how to bring very similar discussuons into one place 😃

**Comment by oandregal** [comment]
:wave: Saw the ping in https://github.com/WordPress/gutenberg/discussions/36079 and here. Thought I'd share the edge cases we've run into that the current block supports mechanism struggles to incorporate. It's mostly captured in this issue and the related links https://github.com/WordPress/gutenberg/issues/33255

TLDR is that the current block supports targets a single area for styling (you can't have two different background colors, for example). While https://github.com/WordPress/gutenberg/issues/28913 allow us to tweak that area via the CSS selector so we can address parts of the block other than the wrapper, it's still a single area.

**Reply by luisherranz** [reply]
Hey André, thanks for chiming in 🙂👋

> While #28913 allow us to tweak that area via the CSS selector so we can address parts of the block other than the wrapper, it's still a single area.

You mean that is a single area when you use `_experimentalSelector`, because neither [using property selectors as Ben suggested here](https://github.com/WordPress/gutenberg/issues/28913#issuecomment-785211181) or [expanding the elements API as Aaron suggested here](https://github.com/WordPress/gutenberg/issues/32417#issuecomment-895728203) have been implemented yet, is that right?

If so, the current way to spread the different support classes/styles is by using the different `experimentalSkipSerialization` flags and their `useXxxProps()` counterparts, like in [the Button block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/edit.js). Is that right?

----

By the way, in #33255 you mention the possibility of having custom styles and controls in the future:

> There's some progress for the new UI for global styles at #32923 so work on allowing custom styles and controls may need to wait a bit.

So, is there a plan to allow custom styles? In other words, are there plans to make block supports extensible by plugins?

**Reply by oandregal** [reply]
There are a few interrelated concepts here:

- How styles & classes are serialized to the block. If a block uses the hooks mechanism, the values chosen by the user in the respective UI controls will be serialized to the wrapper automatically. You can "disengage" the automatic serialization mechanism and do-it-yourself via `__skipSerialization`. This is what the button block does.

- By "single area for styles" I mean that you can only have a single style type per block with the current block supports mechanism. For example, you can only have a single background color per block. So the question is, how do we allow more "background colors" per block?

One underdeveloped aspect of the current API is the "elements" of a block. While this works at the "global styles" level (theme.json and the global styles sidebar, see https://github.com/WordPress/gutenberg/issues/34574 and [this screenshot](https://user-images.githubusercontent.com/548849/132188024-1536a6c4-14f2-4f20-89c3-5e4d03de5c2f.png) in particular), there's no similar concept yet for the "local styles" (block sidebar). Can elements be declared by a block and so its sidebar gains new areas for styling?

That's a promising path in my view. At the moment, we sort of have the footsteps for this: blocks can declare that they support link color and users will be able to tweak the _text color_ of the link element. However, they should be able to do more: change the _background color_ of the link element, change the _typography_ of the link element, etc. If we add more elements (buttons, header, footer, etc.) that blocks can support, we're increasing the "style areas" of the blocks, if that makes sense :) 

> So, is there a plan to allow custom styles? In other words, are there plans to make block supports extensible by plugins?

Blocks can always have custom style controls by hooking into the edit function, etc. Though the issue with custom controls for the block sidebar is that we don't have a way to register them in the global styles sidebar yet. Folks shared concerns in the past against allowing custom controls in the global styles sidebar, as the main idea behind theme.json & the GS sidebar is to develop a system that works across themes. My personal view is that we should ship the global styles UI in 5.9 and listen to what the community says.

**Reply by luisherranz** [reply]
These are very interesting thoughts, thanks for sharing André.

> Can elements be declared by a block and so its sidebar gains new areas for styling?

So the path you mention would be to let blocks express their styling capabilities using elements as the base unit, instead of block supports.

Instead of specifying where each block support should sit in the markup (_oversimplified_):

```js
const Button = ({ attributes }) => {
  const colorProps = useColorProps(attributes);

  return (
    <div>
      <a {...colorProps}>{attributes.content}</a>
    </div>
  );
};
```

Specify which part of the markup represents each styling element:

```js
const Button = ({ attributes }) => {
  const linkProps = uselinkElementProps(attributes);

  return (
    <div>
      <a {...linkProps}>{attributes.content}</a>
    </div>
  );
};
```

In one block, the `link` element can be the `<a>` tags, in other `<button>` tags, and so on...

> If we add more elements (buttons, header, footer, etc.) that blocks can support, we're increasing the "style areas" of the blocks

When you refer to elements like header or footer, you mean semantic parts of a block, not HTML tags, don't you? Like for example, bear with me:

```json
// block.json
{
  "elements": ["header", "main"]
}
```

Or even:

```json
// block.json
{
  "elements": {
    "header": {
      "typography": {
        "fontSize": true
      },
      "color": {
        "link": true,
        "background": true
      }
    },
    "main": {
      "spacing": {
        "padding": true
      }
    }
  }
}
```

```js
const HeroBlock = ({ attributes }) => {
  const header = useHeaderElementProps(attributes);
  const main = useMainElementProps(attributes);

  return (
    <div>
      <div {...header}>
        <h1>{attributes.title}</h1>
        <img src={attributes.src} />
      </div>
      <p {...main}>{attributes.content}</p>
    </div>
  );
};
```

Well, sorry because that example is a bit rough, and it has a few problems, like:

- What happens with elements inside `RichText` generated content, like links?
- What happens if you need to spread supports from one element between different HTML nodes?

So maybe using selectors would be a better way to handle the spread, as Ben proposed. Something like:

```json
// block.json
{
  "supports": {
    "header": {
      // General selector for the header element.
      "selector": ".wp-block-hero__header",
      "typography": {
        "fontSize": true
      },
      "spacing": {
        "padding": {
          // Specific selector for the padding styling of the header element.
          "selector": ".wp-block-hero__header h1"
        },
        // This will fallback to the general selector.
        "margin": true
      }
    },
    "main": {
      // General selector for the main element. Optional.
      "selector": ".wp-block-hero__main",
      "spacing": {
        "padding": true
      }
    },
    // With no selector, this will fallback to the block root.
    "color": {
      "text": true,
      "link": true
    }
  }
}
```

```js
const HeroBlock = ({ attributes }) => (
  <div {...useBlockProps()}>
    <div class="wp-block-hero__header">
      <h1>{attributes.title}</h1>
      <img src={attributes.src} />
    </div>
    <p class="wp-block-hero__main">{attributes.content}</p>
  </div>
);
```

Anyway, examples are not important, I'm just trying to see if I start understanding the current status and plans forward of each Gutenberg API. Let me know if this is the idea 🙂

> So the question is, how do we allow more "background colors" per block?

Are there any plans for this or is it still in the making?

**Reply by cbravobernal** [reply]
If this can help with the discussion giving more context.[ Here we have an example of applying different styles to different components](https://github.com/WordPress/gutenberg/pull/36322) both in JS and in PHP.
Feel free to discuss about it.
